File: modules/agent/src/repositories/organization_repository.py
"""Repository layer for organization data access."""

import logging
from datetime import datetime
from typing import List, Optional, Dict, Any

from pydantic import BaseModel, field_validator
from sqlalchemy import select, func
from sqlalchemy.orm import selectinload

from lib.db import async_get_session
from lib.validators import validate_phone
from models.Account import Account
from models.Industry import Account_Industry
from models.Organization import Organization
from models.OrganizationTechnology import OrganizationTechnology

logger = logging.getLogger(__name__)


# Pydantic models

class OrganizationCreate(BaseModel):
    name: str
    website: Optional[str] = None
    phone: Optional[str] = None
    employee_count: Optional[int] = None
    employee_count_range_id: Optional[int] = None
    funding_stage_id: Optional[int] = None
    description: Optional[str] = None
    account_id: Optional[int] = None
    industry_ids: Optional[List[int]] = None
    project_ids: Optional[List[int]] = None
    revenue_range_id: Optional[int] = None
    founding_year: Optional[int] = None
    headquarters_city: Optional[str] = None
    headquarters_state: Optional[str] = None
    headquarters_country: Optional[str] = None
    company_type: Optional[str] = None
    linkedin_url: Optional[str] = None
    crunchbase_url: Optional[str] = None

    @field_validator("phone")
    @classmethod
    def validate_phone_format(cls, v):
        return validate_phone(v)

    @field_validator("founding_year")
    @classmethod
    def validate_founding_year(cls, v):
        if v is None:
            return v
        current_year = datetime.now().year
        if v < 1800 or v > current_year:
            raise ValueError(f"founding_year must be between 1800 and {current_year}")
        return v


class OrganizationUpdate(BaseModel):
    name: Optional[str] = None
    website: Optional[str] = None
    phone: Optional[str] = None
    employee_count: Optional[int] = None
    employee_count_range_id: Optional[int] = None
    funding_stage_id: Optional[int] = None
    description: Optional[str] = None
    industry_ids: Optional[List[int]] = None
    project_ids: Optional[List[int]] = None
    revenue_range_id: Optional[int] = None
    founding_year: Optional[int] = None
    headquarters_city: Optional[str] = None
    headquarters_state: Optional[str] = None
    headquarters_country: Optional[str] = None
    company_type: Optional[str] = None
    linkedin_url: Optional[str] = None
    crunchbase_url: Optional[str] = None

    @field_validator("phone")
    @classmethod
    def validate_phone_format(cls, v):
        return validate_phone(v)

    @field_validator("founding_year")
    @classmethod
    def validate_founding_year(cls, v):
        if v is None:
            return v
        current_year = datetime.now().year
        if v < 1800 or v > current_year:
            raise ValueError(f"founding_year must be between 1800 and {current_year}")
        return v


class OrgContactCreate(BaseModel):
    individual_id: Optional[int] = None
    first_name: str
    last_name: str
    title: Optional[str] = None
    department: Optional[str] = None
    role: Optional[str] = None
    email: Optional[str] = None
    phone: Optional[str] = None
    is_primary: bool = False
    notes: Optional[str] = None

    @field_validator("phone")
    @classmethod
    def validate_phone_format(cls, v):
        return validate_phone(v)


class OrgContactUpdate(BaseModel):
    first_name: Optional[str] = None
    last_name: Optional[str] = None
    title: Optional[str] = None
    department: Optional[str] = None
    role: Optional[str] = None
    email: Optional[str] = None
    phone: Optional[str] = None
    is_primary: Optional[bool] = None
    notes: Optional[str] = None

    @field_validator("phone")
    @classmethod
    def validate_phone_format(cls, v):
        return validate_phone(v)


class OrgTechnologyCreate(BaseModel):
    technology_id: int
    source: Optional[str] = None
    confidence: Optional[float] = None


class FundingRoundCreate(BaseModel):
    round_type: str
    amount: Optional[int] = None
    announced_date: Optional[str] = None
    lead_investor: Optional[str] = None
    source_url: Optional[str] = None


class SignalCreate(BaseModel):
    signal_type: str
    headline: str
    description: Optional[str] = None
    signal_date: Optional[str] = None
    source: Optional[str] = None
    source_url: Optional[str] = None
    sentiment: Optional[str] = None
    relevance_score: Optional[float] = None


async def find_all_organizations(
    tenant_id: Optional[int] = None,
    page: int = 1,
    page_size: int = 50,
    order_by: str = "updated_at",
    order: str = "DESC",
    search: Optional[str] = None,
) -> tuple[List[Organization], int]:
    """Find organizations with pagination, filtering, and sorting at database level.
    
    Optimized for list view - only loads essential relationships:
    - Account industries (for industry display)
    - Employee count range (for employees display)
    - Funding stage (for funding display)
    """
    async with async_get_session() as session:
        # Base query with minimal relationships needed for list view
        stmt = select(Organization).options(
            selectinload(Organization.account).selectinload(Account.industries),
            selectinload(Organization.employee_count_range),
            selectinload(Organization.funding_stage),
        )
        
        if tenant_id is not None:
            stmt = stmt.join(Account, Organization.account_id == Account.id).where(Account.tenant_id == tenant_id)

        # Apply search filter at DB level
        if search and search.strip():
            search_lower = f"%{search.strip().lower()}%"
            stmt = stmt.where(
                func.lower(Organization.name).like(search_lower)
            )

        # Get total count before pagination
        count_stmt = select(func.count()).select_from(stmt.subquery())
        total = (await session.execute(count_stmt)).scalar() or 0

        # Apply sorting at DB level
        sort_map = {
            "name": Organization.name,
            "updated_at": Organization.updated_at,
            "created_at": Organization.created_at,
            "description": Organization.description,
            "website": Organization.website,
            "funding_stage": Organization.funding_stage_id,
            "employee_count_range": Organization.employee_count_range_id,
            "industries": Organization.account_id,  # Sort by account_id as proxy for industries
        }
        sort_column = sort_map.get(order_by, Organization.updated_at)
        if order.upper() == "ASC":
            stmt = stmt.order_by(sort_column.asc())
        else:
            stmt = stmt.order_by(sort_column.desc())

        # Apply pagination
        offset = (page - 1) * page_size
        stmt = stmt.offset(offset).limit(page_size)

        result = await session.execute(stmt)
        return list(result.scalars().all()), total


async def find_organization_by_id(org_id: int) -> Optional[Organization]:
    """Find organization by ID with all relationships for detail view."""
    async with async_get_session() as session:
        stmt = (
            select(Organization)
            .options(
                selectinload(Organization.account).selectinload(Account.industries),
                selectinload(Organization.account).selectinload(Account.projects),
                selectinload(Organization.employee_count_range),
                selectinload(Organization.funding_stage),
                selectinload(Organization.revenue_range),
                selectinload(Organization.technologies).selectinload(OrganizationTechnology.technology),
                selectinload(Organization.funding_rounds),
                selectinload(Organization.signals),
            )
            .where(Organization.id == org_id)
        )
        result = await session.execute(stmt)
        return result.scalar_one_or_none()


async def find_organization_by_name(name: str, tenant_id: Optional[int] = None) -> Optional[Organization]:
    """Find organization by name (case-insensitive), optionally scoped to tenant (through Account)."""
    name = name.strip() if name else name
    async with async_get_session() as session:
        stmt = select(Organization).where(func.lower(Organization.name) == func.lower(name))
        if tenant_id is not None:
            stmt = stmt.join(Account, Organization.account_id == Account.id).where(Account.tenant_id == tenant_id)
        result = await session.execute(stmt)
        return result.scalar_one_or_none()


async def create_organization(data: Dict[str, Any]) -> Organization:
    """Create a new organization."""
    async with async_get_session() as session:
        try:
            org = Organization(**data)
            session.add(org)
            await session.commit()
            # Re-fetch with relationships loaded
            stmt = (
                select(Organization)
                .options(
                    selectinload(Organization.account).selectinload(Account.industries),
                    selectinload(Organization.employee_count_range),
                    selectinload(Organization.funding_stage),
                    selectinload(Organization.revenue_range),
                )
                .where(Organization.id == org.id)
            )
            result = await session.execute(stmt)
            return result.scalar_one()
        except Exception as e:
            await session.rollback()
            logger.error(f"Failed to create organization: {e}")
            raise


async def get_or_create_organization(
    name: str,
    tenant_id: Optional[int] = None,
    industry_ids: Optional[List[int]] = None
) -> Organization:
    """Get or create organization by name, optionally linking industries to the account."""
    name = name.strip() if name else name
    existing = await find_organization_by_name(name, tenant_id)
    if existing:
        return existing

    # Create account for the organization with tenant_id
    async with async_get_session() as session:
        try:
            account = Account(
                tenant_id=tenant_id,
                name=name
            )
            session.add(account)
            await session.flush()

            org = Organization(
                account_id=account.id,
                name=name
            )
            session.add(org)
            await session.flush()

            # Link industries to the account if provided
            if industry_ids:
                for industry_id in industry_ids:
                    stmt = Account_Industry.insert().values(
                        account_id=account.id,
                        industry_id=industry_id
                    )
                    await session.execute(stmt)

            await session.commit()
            await session.refresh(org)
            return org
        except Exception as e:
            await session.rollback()
            logger.error(f"Failed to create organization: {e}")
            raise


async def update_organization(org_id: int, data: Dict[str, Any]) -> Optional[Organization]:
    """Update an existing organization."""
    async with async_get_session() as session:
        try:
            org = await session.get(Organization, org_id)
            if not org:
                return None

            for key, value in data.items():
                if hasattr(org, key):
                    setattr(org, key, value)

            await session.commit()
            # Re-fetch with relationships loaded
            stmt = (
                select(Organization)
                .options(
                    selectinload(Organization.account).selectinload(Account.industries),
                    selectinload(Organization.employee_count_range),
                    selectinload(Organization.funding_stage),
                    selectinload(Organization.revenue_range),
                )
                .where(Organization.id == org_id)
            )
            result = await session.execute(stmt)
            return result.scalar_one()
        except Exception as e:
            await session.rollback()
            logger.error(f"Failed to update organization: {e}")
            raise


async def delete_organization(org_id: int) -> bool:
    """Delete an organization by ID."""
    async with async_get_session() as session:
        try:
            org = await session.get(Organization, org_id)
            if not org:
                return False
            await session.delete(org)
            await session.commit()
            return True
        except Exception as e:
            await session.rollback()
            logger.error(f"Failed to delete organization: {e}")
            raise


async def search_organizations(query: str, tenant_id: Optional[int] = None) -> List[Organization]:
    """Search organizations by name (case-insensitive partial match), filtered by tenant (through Account)."""
    async with async_get_session() as session:
        stmt = select(Organization).options(
            selectinload(Organization.account).selectinload(Account.industries),
            selectinload(Organization.employee_count_range),
            selectinload(Organization.funding_stage),
            selectinload(Organization.revenue_range),
        ).where(
            func.lower(Organization.name).contains(func.lower(query))
        ).order_by(Organization.name)
        if tenant_id is not None:
            stmt = stmt.join(Account, Organization.account_id == Account.id).where(Account.tenant_id == tenant_id)
        result = await session.execute(stmt)
        return list(result.scalars().all())

File: modules/agent-go/internal/repositories/organization_repository.go
package repositories

import (
	"time"

	"github.com/pina-colada-co/agent-go/internal/models"
	"github.com/shopspring/decimal"
	"gorm.io/gorm"
)

// OrganizationRepository handles organization data access
type OrganizationRepository struct {
	db *gorm.DB
}

// NewOrganizationRepository creates a new organization repository
func NewOrganizationRepository(db *gorm.DB) *OrganizationRepository {
	return &OrganizationRepository{db: db}
}

// FindAll returns paginated organizations
func (r *OrganizationRepository) FindAll(params PaginationParams, tenantID *int64) (*PaginatedResult[models.Organization], error) {
	var orgs []models.Organization
	var totalCount int64

	query := r.db.Model(&models.Organization{}).
		Preload("Account").
		Preload("Account.Industries").
		Preload("Technologies").
		Preload("EmployeeCountRange").
		Preload("FundingStage")

	// Filter by tenant (through Account)
	if tenantID != nil {
		query = query.Joins(`INNER JOIN "Account" ON "Account".id = "Organization".account_id`).
			Where(`"Account".tenant_id = ?`, *tenantID)
	}

	// Apply search
	if params.Search != "" {
		searchTerm := "%" + params.Search + "%"
		query = query.Where(`LOWER("Organization".name) LIKE LOWER(?)`, searchTerm)
	}

	// Count total
	if err := query.Count(&totalCount).Error; err != nil {
		return nil, err
	}

	// Apply pagination
	query = ApplyPagination(query, params)

	if err := query.Find(&orgs).Error; err != nil {
		return nil, err
	}

	return &PaginatedResult[models.Organization]{
		Items:      orgs,
		TotalCount: totalCount,
		Page:       params.Page,
		PageSize:   params.PageSize,
		TotalPages: CalculateTotalPages(totalCount, params.PageSize),
	}, nil
}

// FindByID returns an organization by ID
func (r *OrganizationRepository) FindByID(id int64) (*models.Organization, error) {
	var org models.Organization
	err := r.db.
		Preload("Account").
		Preload("Account.Industries").
		Preload("Technologies").
		Preload("FundingRounds").
		First(&org, id).Error

	if err != nil {
		if err == gorm.ErrRecordNotFound {
			return nil, nil
		}
		return nil, err
	}
	return &org, nil
}

// FindByName searches organizations by name
func (r *OrganizationRepository) FindByName(name string, tenantID *int64, limit int) ([]models.Organization, error) {
	var orgs []models.Organization

	query := r.db.Model(&models.Organization{}).
		Preload("Account").
		Where(`LOWER("Organization".name) LIKE LOWER(?)`, "%"+name+"%")

	if tenantID != nil {
		query = query.Joins(`INNER JOIN "Account" ON "Account".id = "Organization".account_id`).
			Where(`"Account".tenant_id = ?`, *tenantID)
	}

	err := query.Limit(limit).Find(&orgs).Error
	return orgs, err
}

// FindOrCreateByName finds an organization by exact name or creates it with an Account
func (r *OrganizationRepository) FindOrCreateByName(name string, tenantID *int64, userID int64) (*models.Organization, error) {
	var org models.Organization
	err := r.db.Preload("Account").
		Where(`LOWER("Organization".name) = LOWER(?)`, name).
		First(&org).Error

	if err == nil {
		return &org, nil
	}
	if err != gorm.ErrRecordNotFound {
		return nil, err
	}

	// Create Account first
	account := &models.Account{
		TenantID:  tenantID,
		Name:      name,
		CreatedBy: userID,
		UpdatedBy: userID,
	}
	if err := r.db.Create(account).Error; err != nil {
		return nil, err
	}

	// Create Organization linked to Account
	newOrg := &models.Organization{
		AccountID: &account.ID,
		Name:      name,
		CreatedBy: userID,
		UpdatedBy: userID,
	}
	if err := r.db.Create(newOrg).Error; err != nil {
		return nil, err
	}

	newOrg.Account = account
	return newOrg, nil
}

// Create creates a new organization
func (r *OrganizationRepository) Create(org *models.Organization) error {
	return r.db.Create(org).Error
}

// Update updates an organization
func (r *OrganizationRepository) Update(org *models.Organization, updates map[string]interface{}) error {
	return r.db.Model(&models.Organization{ID: org.ID}).Updates(updates).Error
}

// Delete deletes an organization by ID
func (r *OrganizationRepository) Delete(id int64) error {
	return r.db.Delete(&models.Organization{}, id).Error
}

// UpdateAccountIndustries replaces an account's industries
func (r *OrganizationRepository) UpdateAccountIndustries(accountID int64, industryIDs []int64) error {
	// Delete existing
	if err := r.db.Where("account_id = ?", accountID).Delete(&models.AccountIndustry{}).Error; err != nil {
		return err
	}
	// Insert new
	for _, indID := range industryIDs {
		ai := models.AccountIndustry{AccountID: accountID, IndustryID: indID}
		if err := r.db.Create(&ai).Error; err != nil {
			return err
		}
	}
	return nil
}

// OrgContactInput holds data for creating a contact for an org
type OrgContactInput struct {
	FirstName string
	LastName  string
	Email     *string
	Phone     *string
	Title     *string
	IsPrimary bool
}

// CreateContactForAccount creates a contact and links it to an account
func (r *OrganizationRepository) CreateContactForAccount(accountID int64, input OrgContactInput, userID int64) (*models.Contact, error) {
	contact := &models.Contact{
		FirstName: input.FirstName,
		LastName:  input.LastName,
		Email:     input.Email,
		Phone:     input.Phone,
		Title:     input.Title,
		CreatedBy: userID,
		UpdatedBy: userID,
	}

	if err := r.db.Create(contact).Error; err != nil {
		return nil, err
	}

	// Link to account
	link := &models.ContactAccount{
		ContactID: contact.ID,
		AccountID: accountID,
		IsPrimary: input.IsPrimary,
	}
	if err := r.db.Create(link).Error; err != nil {
		return nil, err
	}

	return contact, nil
}

// GetContactsForAccount returns all contacts for an account
func (r *OrganizationRepository) GetContactsForAccount(accountID int64) ([]models.Contact, error) {
	var contacts []models.Contact
	err := r.db.Table(`"Contact"`).
		Joins(`INNER JOIN "Contact_Account" ON "Contact".id = "Contact_Account".contact_id`).
		Where(`"Contact_Account".account_id = ?`, accountID).
		Find(&contacts).Error
	return contacts, err
}

// UpdateContact updates a contact
func (r *OrganizationRepository) UpdateContact(contactID int64, updates map[string]interface{}) error {
	return r.db.Model(&models.Contact{}).Where("id = ?", contactID).Updates(updates).Error
}

// DeleteContactFromAccount removes a contact's link to an account
func (r *OrganizationRepository) DeleteContactFromAccount(accountID int64, contactID int64) error {
	return r.db.Where("account_id = ? AND contact_id = ?", accountID, contactID).
		Delete(&models.ContactAccount{}).Error
}

// GetTechnologiesForOrg returns all technologies for an organization
func (r *OrganizationRepository) GetTechnologiesForOrg(orgID int64) ([]models.OrganizationTechnology, error) {
	var techs []models.OrganizationTechnology
	err := r.db.Where("organization_id = ?", orgID).Find(&techs).Error
	return techs, err
}

// AddTechnologyToOrg adds a technology to an organization
func (r *OrganizationRepository) AddTechnologyToOrg(orgID int64, techID int64, source *string, confidence *float64) error {
	tech := &models.OrganizationTechnology{
		OrganizationID: orgID,
		TechnologyID:   techID,
		Source:         source,
	}
	if confidence != nil {
		dec := decimal.NewFromFloat(*confidence)
		tech.Confidence = &dec
	}
	return r.db.Create(tech).Error
}

// RemoveTechnologyFromOrg removes a technology from an organization
func (r *OrganizationRepository) RemoveTechnologyFromOrg(orgID int64, techID int64) error {
	return r.db.Where("organization_id = ? AND technology_id = ?", orgID, techID).
		Delete(&models.OrganizationTechnology{}).Error
}

// GetFundingRoundsForOrg returns all funding rounds for an organization
func (r *OrganizationRepository) GetFundingRoundsForOrg(orgID int64) ([]models.FundingRound, error) {
	var rounds []models.FundingRound
	err := r.db.Where("organization_id = ?", orgID).Order("announced_date DESC").Find(&rounds).Error
	return rounds, err
}

// CreateFundingRound creates a funding round for an organization
func (r *OrganizationRepository) CreateFundingRound(round *models.FundingRound) error {
	return r.db.Create(round).Error
}

// DeleteFundingRound deletes a funding round
func (r *OrganizationRepository) DeleteFundingRound(roundID int64) error {
	return r.db.Delete(&models.FundingRound{}, roundID).Error
}

// GetSignalsForAccount returns signals for an account
func (r *OrganizationRepository) GetSignalsForAccount(accountID int64, signalType *string, limit int) ([]models.Signal, error) {
	var signals []models.Signal
	query := r.db.Where("account_id = ?", accountID)
	if signalType != nil {
		query = query.Where("signal_type = ?", *signalType)
	}
	err := query.Order("created_at DESC").Limit(limit).Find(&signals).Error
	return signals, err
}

// CreateSignal creates a signal for an account
func (r *OrganizationRepository) CreateSignal(signal *models.Signal) error {
	return r.db.Create(signal).Error
}

// DeleteSignal deletes a signal
func (r *OrganizationRepository) DeleteSignal(signalID int64) error {
	return r.db.Delete(&models.Signal{}, signalID).Error
}

// FundingRoundInput holds data for creating a funding round
type FundingRoundInput struct {
	OrganizationID int64
	RoundType      string
	Amount         *int64
	AnnouncedDate  *string
	LeadInvestor   *string
	SourceURL      *string
}

// CreateFundingRoundFromInput creates a funding round from input
func (r *OrganizationRepository) CreateFundingRoundFromInput(input *FundingRoundInput) (*models.FundingRound, error) {
	round := &models.FundingRound{
		OrganizationID: input.OrganizationID,
		RoundType:      input.RoundType,
		Amount:         input.Amount,
		LeadInvestor:   input.LeadInvestor,
		SourceURL:      input.SourceURL,
	}
	if input.AnnouncedDate != nil {
		if t, err := parseDate(*input.AnnouncedDate); err == nil {
			round.AnnouncedDate = &t
		}
	}
	if err := r.db.Create(round).Error; err != nil {
		return nil, err
	}
	return round, nil
}

// CreateSignalForAccount creates a signal for an account
func (r *OrganizationRepository) CreateSignalForAccount(accountID int64, input SignalCreateInput) (*models.Signal, error) {
	signal := &models.Signal{
		AccountID:      accountID,
		SignalType:     input.SignalType,
		Headline:       input.Headline,
		Description:    input.Description,
		Source:         input.Source,
		SourceURL:      input.SourceURL,
		Sentiment:      input.Sentiment,
		SignalDate:     input.SignalDate,
		RelevanceScore: input.RelevanceScore,
	}
	if err := r.db.Create(signal).Error; err != nil {
		return nil, err
	}
	return signal, nil
}

// CreateAccount creates an account
func (r *OrganizationRepository) CreateAccount(name string, tenantID *int64, userID int64) (*models.Account, error) {
	account := &models.Account{
		TenantID:  tenantID,
		Name:      name,
		CreatedBy: userID,
		UpdatedBy: userID,
	}
	if err := r.db.Create(account).Error; err != nil {
		return nil, err
	}
	return account, nil
}

// OrganizationCreateInput holds data for creating an organization
type OrganizationCreateInput struct {
	AccountID            int64
	Name                 string
	Website              *string
	Phone                *string
	EmployeeCount        *int
	EmployeeCountRangeID *int64
	FundingStageID       *int64
	Description          *string
	RevenueRangeID       *int64
	FoundingYear         *int
	HeadquartersCity     *string
	HeadquartersState    *string
	HeadquartersCountry  *string
	CompanyType          *string
	LinkedInURL          *string
	CrunchbaseURL        *string
	UserID               int64
}

// CreateOrganization creates an organization
func (r *OrganizationRepository) CreateOrganization(input OrganizationCreateInput) (*models.Organization, error) {
	org := &models.Organization{
		AccountID:            &input.AccountID,
		Name:                 input.Name,
		Website:              input.Website,
		Phone:                input.Phone,
		EmployeeCount:        input.EmployeeCount,
		EmployeeCountRangeID: input.EmployeeCountRangeID,
		FundingStageID:       input.FundingStageID,
		Description:          input.Description,
		RevenueRangeID:       input.RevenueRangeID,
		FoundingYear:         input.FoundingYear,
		HeadquartersCity:     input.HeadquartersCity,
		HeadquartersState:    input.HeadquartersState,
		HeadquartersCountry:  input.HeadquartersCountry,
		CompanyType:          input.CompanyType,
		LinkedInURL:          input.LinkedInURL,
		CrunchbaseURL:        input.CrunchbaseURL,
		CreatedBy:            input.UserID,
		UpdatedBy:            input.UserID,
	}
	if err := r.db.Create(org).Error; err != nil {
		return nil, err
	}
	return org, nil
}

func parseDate(s string) (time.Time, error) {
	return time.Parse("2006-01-02", s)
}
